sample-code/FSM/src/lang/fsm/WFR.rsc

```diff
@@ -5,43 +5,46 @@ import List;
 import Set;
 import lang::fsm::AbstractSyntax; 
 
-data Error = moreThanOneStartState()
-           | noFinalState()
-           | noStartState()
-           | ambiguousTransitions(list[Transition] transitions);
+data Error = noUniqueStartState(list[State] initialStates)
+           | unresolvableTargetState()
+           | ambiguousTransitions(list[Transition] transitions)
+           | unreachableStates();
 
 
-bool singleInitialState(StateMachine sm) = 1 == size([s | State s <- sm.states, startElement(s) == 1]);
+list[Error] singleInitialState(StateMachine sm) = 1 == size(startStates(sm)) ? [] :  [noUniqueStartState(startStates(sm))];
 
-bool resolvableTargetState(StateMachine sm) = toSet([t.target | Transition t <- sm.transitions]) <= toSet(sm.states);
+bool resolvableTargetState(StateMachine sm) = toSet([t.target | Transition t <- sm.transitions]) == toSet(sm.states);
 
 bool distinctStates(StateMachine sm) =  size(toSet([s.name | State s <- sm.states])) == size([s.name | State s <- sm.states]);
 
-bool deterministicTransitions(StateMachine sm) = showAmbiguos(showListTransitions(sm))==[];
+list[Error] deterministicTransitions(StateMachine sm) = isEmpty(showAmbiguos(showListTransitions(sm))) ? [] :[ambiguousTransitions(showAmbiguos(showListTransitions(sm)))];
+
+public set[State] visited = {};
+list[State] res = []; 
 
 bool reachableState(StateMachine sm) {
-	list[State] initialState = [s | State s <- sm.states, startElement(s) == 1];
-	
-	return toSet(reachables(sm, initialState[0])) == toSet(sm.states);
+	list[State] initialState = startStates(sm);
+	reachables(sm, initialState[0]);
+	return toSet(res) == toSet(sm.states);
 }
 
 //Help functions to identify WFR: -------------------------------------------------------------------------------------
 
-int startElement(startState(_)) = 1;
-int startElement(state(_)) = 0;
+
+list[State] startStates(StateMachine sm) = [startState(n) | startState(n) <- sm.states];
 
 public list[Transition] showAmbiguos (list[Transition] l) = isEmpty(l) ? [] :findAmbiguos(head(l),tail(l))+showAmbiguos(tail(l));
 list[Transition] findAmbiguos(Transition t,list[Transition] lt) = [X | Transition X <- lt,(X.source==t.source) && (X.event==t.event)];
 
 public list[Transition] showListTransitions(StateMachine sm) = ([t | /Transition t <- sm]);
 
-public set[State] visited = {};
-list[State] reachables(StateMachine sm, State s){
-	if( s in visited){
-		return [];
-	} else {
+public void reachables(StateMachine sm, State s){
```

Comment:
I have a suggestion to refactor this, but I don't know how we can translate the `map`, `flatten` and `reduce` functions to Rascal, but here's the *idea*:

```
list[State] reachables(State s, list[State] reachablesStates) {
  list[Transition] transitions = getTransitionsFromState(s)

  list[State] newReachablesStates = transitions.reduce((acc, transition) => {
    if (!(transition.source in acc)) {
      acc += transition.source
    }
  }, reachablesStates)
  newReachablesStates += s

  return flatten(newReachablesStates.map((s) => reachables(s, newReachablesStates)))
}

list[States] reachableStatesFromInitialState = toSet(reachables(sm.initialState, []))
```